File: scripts/analysis/vtc_validation/run_layer1_controlled_recovery.py
"""Pure-Python Layer 1 runner for the frozen VTC validation contract."""

from __future__ import annotations

import argparse
import csv
import json
import math
from collections.abc import Iterator
from pathlib import Path
from statistics import median
from typing import Any

from vtc_validation_common import (
    EstimatorConfig,
    MatchingTolerances,
    ValidationCase,
    estimate_joint,
    estimator_from_contract,
    inject_observations,
    load_observations,
    match_injected,
    prepare_case,
    tolerances_from_contract,
)


ROOT = Path(__file__).resolve().parents[3]
DEFAULT_CONTRACT = ROOT / "docs/vtc2027_spring/evidence/validation_v1/validation_contract.json"

FIELDS = [
    "trial_id", "layer", "scene_id", "prn_label", "environment",
    "center_window_id", "source_interval_start_zero_based",
    "source_interval_end_zero_based", "excess_delay_truth_samples",
    "relative_doppler_truth_hz", "relative_power_truth_db", "phase_truth_rad",
    "selected_order", "joint_valid", "injected_match", "delay_error_samples",
    "doppler_error_hz", "power_error_db", "match_cost", "joint_rss",
    "joint_bic", "snapshot_wins", "failure_reason",
]


def enumerate_trials(contract: dict[str, Any]) -> Iterator[dict[str, Any]]:
    ordinal = 0
    layer = contract["layer1"]
    for case in layer["cases"]:
        for delay in layer["excess_delay_samples"]:
            for doppler in layer["relative_doppler_hz"]:
                for power in layer["relative_power_db"]:
                    for phase in layer["relative_phase_rad"]:
                        ordinal += 1
                        yield {
                            "trial_id": f"L1_{ordinal:04d}",
                            "case": case,
                            "delay": float(delay),
                            "doppler": float(doppler),
                            "power": float(power),
                            "phase": float(phase),
                        }


def _base_row(case: ValidationCase, trial: dict[str, Any]) -> dict[str, Any]:
    starts = [int(snapshot["sample_start_zero_based"]) for snapshot in case.snapshots]
    return {
        "trial_id": trial["trial_id"],
        "layer": "Layer1_Controlled",
        "scene_id": case.scene_id,
        "prn_label": case.prn_label,
        "environment": case.environment,
        "center_window_id": case.center_window_id,
        "source_interval_start_zero_based": min(starts),
        "source_interval_end_zero_based": max(starts) + case.sample_count - 1,
        "excess_delay_truth_samples": trial["delay"],
        "relative_doppler_truth_hz": trial["doppler"],
        "relative_power_truth_db": trial["power"],
        "phase_truth_rad": trial["phase"],
        "selected_order": math.nan,
        "joint_valid": False,
        "injected_match": False,
        "delay_error_samples": math.nan,
        "doppler_error_hz": math.nan,
        "power_error_db": math.nan,
        "match_cost": math.nan,
        "joint_rss": math.nan,
        "joint_bic": math.nan,
        "snapshot_wins": 0,
        "failure_reason": "NOT_RUN",
    }


def execute_trial(
    case: ValidationCase,
    base_observations: list,
    trial: dict[str, Any],
    estimator: EstimatorConfig,
    tolerances: MatchingTolerances,
) -> dict[str, Any]:
    row = _base_row(case, trial)
    observations = inject_observations(
        base_observations, case, trial["delay"], trial["doppler"],
        trial["power"], trial["phase"],
    )
    result = estimate_joint(
        observations, case.direct_path, case.contexts, case.doppler_bound_hz, estimator
    )
    match = match_injected(
        result.selected, trial["delay"], trial["doppler"], trial["power"], tolerances
    )
    row.update(
        selected_order=result.selected_order,
        joint_valid=result.joint_valid,
        injected_match=match.found,
        delay_error_samples=match.delay_error_samples,
        doppler_error_hz=match.doppler_error_hz,
        power_error_db=match.power_error_db,
        match_cost=match.cost,
        joint_rss=result.joint_rss,
        joint_bic=result.joint_bic,
        snapshot_wins=result.snapshot_wins,
    )
    if result.selected_order < 2:
        row["failure_reason"] = "NO_SECONDARY_MODEL_SELECTED"
    elif not result.joint_valid:
        row["failure_reason"] = "SELECTED_MODEL_INVALID"
    elif not match.found:
        row["failure_reason"] = "INJECTED_PATH_OUTSIDE_TOLERANCE"
    else:
        row["failure_reason"] = "PASS"
    return row


def _write_rows(path: Path, rows: list[dict[str, Any]], fields: list[str]) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=fields)
        writer.writeheader()
        writer.writerows(rows)


def _summary(rows: list[dict[str, Any]], powers: list[float]) -> list[dict[str, Any]]:
    output = []
    for power in powers:
        subset = [row for row in rows if float(row["relative_power_truth_db"]) == float(power)]
        finite = [
            row for row in subset
            if all(math.isfinite(float(row[key])) for key in (
                "delay_error_samples", "doppler_error_hz", "power_error_db"
            ))
        ]
        output.append({
            "layer": "Layer1_Controlled",
            "relative_power_db": power,
            "trial_count": len(subset),
            "recovery_count": sum(bool(row["injected_match"]) for row in subset),
            "recovery_rate": sum(bool(row["injected_match"]) for row in subset) / max(len(subset), 1),
            "median_abs_delay_error_samples": median(abs(float(row["delay_error_samples"])) for row in finite) if finite else math.nan,
            "median_abs_doppler_error_hz": median(abs(float(row["doppler_error_hz"])) for row in finite) if finite else math.nan,
            "median_abs_power_error_db": median(abs(float(row["power_error_db"])) for row in finite) if finite else math.nan,
            "finite_error_count": len(finite),
        })
    return output


def run(contract_path: Path = DEFAULT_CONTRACT, max_trials: int | None = None) -> int:
    contract = json.loads(contract_path.read_text(encoding="utf-8"))
    if contract["production_execution"] or contract["resume"]:
        raise ValueError("validation contract is not isolated from production/resume")
    output_dir = Path(contract["output_namespace"])
    output_dir.mkdir(parents=True, exist_ok=True)
    estimator = estimator_from_contract(contract)
    tolerances = tolerances_from_contract(contract)
    rows: list[dict[str, Any]] = []
    case_cache: dict[int, tuple[ValidationCase, list]] = {}
    for ordinal, trial in enumerate(enumerate_trials(contract), start=1):
        if max_trials is not None and ordinal > max_trials:
            break
        center = int(trial["case"]["center_window_id"])
        if center not in case_cache:
            case = prepare_case(contract, trial["case"])
            observations, _ = load_observations(case)
            case_cache[center] = (case, observations)
            print(f"Layer 1 loaded {case.scene_id}/{case.prn_label}/{center}", flush=True)
        case, base = case_cache[center]
        try:
            row = execute_trial(case, base, trial, estimator, tolerances)
        except Exception as exc:  # retain every failed trial
            row = _base_row(case, trial)
            row["failure_reason"] = f"PYTHON_ERROR: {type(exc).__name__}: {exc}"
        rows.append(row)
        if ordinal % 4 == 0:
            _write_rows(output_dir / "layer1_controlled_trials.partial.csv", rows, FIELDS)
            print(f"Layer 1 progress {ordinal}/216", flush=True)
    output_name = "layer1_controlled_trials.csv" if max_trials is None else "layer1_controlled_trials.smoke.csv"
    _write_rows(output_dir / output_name, rows, FIELDS)
    if max_trials is None:
        summary = _summary(rows, [float(value) for value in contract["layer1"]["relative_power_db"]])
        _write_rows(output_dir / "layer1_controlled_summary.csv", summary, list(summary[0]))
    return 0


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--contract", type=Path, default=DEFAULT_CONTRACT)
    parser.add_argument("--max-trials", type=int)
    args = parser.parse_args()
    return run(args.contract, args.max_trials)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/analysis/vtc_validation/run_layer1_controlled_recovery.m
function run_layer1_controlled_recovery(contractPath)
%RUN_LAYER1_CONTROLLED_RECOVERY Execute the frozen G18 controlled layer.
% This is validation-only code. It never calls the production wrapper or
% writes under scenes/**/sage_results.

if nargin < 1
    contractPath = fullfile(fileparts(fileparts(fileparts(fileparts(mfilename("fullpath"))))), ...
        "docs", "vtc2027_spring", "evidence", "validation_v1", ...
        "validation_contract.json");
end
contract = jsondecode(fileread(contractPath));
assert(~contract.production_execution && ~contract.resume, ...
    "Validation contract is not isolated from production/resume.");
outDir = string(contract.output_namespace);
if ~isfolder(outDir)
    mkdir(outDir);
end

estimator = vtc_validation_common("estimator", contract);
tolerances = vtc_validation_common("tolerances", contract);
trialRows = repmat(emptyLayer1Row(), 0, 1);
trialOrdinal = 0;

for caseIndex = 1:numel(contract.layer1.cases)
    caseJson = contract.layer1.cases(caseIndex);
    data = vtc_validation_common("prepare_case", contract, caseJson);
    [baseObservations, readInfo] = vtc_validation_common( ...
        "load_observations", data);
    fprintf("Layer 1 case %s/%s center=%d loaded\n", ...
        data.sceneId, data.prnLabel, data.centerWindowId);

    for delay = contract.layer1.excess_delay_samples
        for doppler = contract.layer1.relative_doppler_hz
            for power = contract.layer1.relative_power_db
                for phase = contract.layer1.relative_phase_rad
                    trialOrdinal = trialOrdinal + 1;
                    row = emptyLayer1Row();
                    row.trial_id = string(sprintf("L1_%04d", trialOrdinal));
                    row.layer = "Layer1_Controlled";
                    row.scene_id = data.sceneId;
                    row.prn_label = data.prnLabel;
                    row.environment = data.environment;
                    row.center_window_id = data.centerWindowId;
                    row.source_interval_start_zero_based = min([readInfo.sample_start_zero_based]);
                    row.source_interval_end_zero_based = max([readInfo.sample_start_zero_based]) + 204600 - 1;
                    row.excess_delay_truth_samples = delay;
                    row.relative_doppler_truth_hz = doppler;
                    row.relative_power_truth_db = power;
                    row.phase_truth_rad = phase;
                    try
                        observations = vtc_validation_common( ...
                            "inject_observations", baseObservations, data, ...
                            delay, doppler, power, phase);
                        result = vtc_validation_common( ...
                            "estimate_joint", observations, data, estimator);
                        match = vtc_validation_common( ...
                            "match_injected", result.selected, delay, doppler, power, tolerances);
                        row.selected_order = result.selectedOrder;
                        row.joint_valid = result.jointValid;
                        row.injected_match = match.found;
                        row.delay_error_samples = match.delayErrorSamples;
                        row.doppler_error_hz = match.dopplerErrorHz;
                        row.power_error_db = match.powerErrorDb;
                        row.match_cost = match.cost;
                        row.joint_rss = result.jointRss;
                        row.joint_bic = result.jointBic;
                        row.snapshot_wins = result.snapshotWins;
                        row.failure_reason = classifyFailure(result, match);
                    catch exception
                        row.failure_reason = "MATLAB_ERROR: " + string(exception.message);
                    end
                    trialRows(end + 1, 1) = row; %#ok<AGROW>
                end
            end
        end
    end
end

trialTable = struct2table(trialRows);
writetable(trialTable, fullfile(outDir, "layer1_controlled_trials.csv"));
summaryTable = makeLayer1Summary(trialTable, contract.layer1.relative_power_db);
writetable(summaryTable, fullfile(outDir, "layer1_controlled_summary.csv"));
fprintf("Layer 1 wrote %d trials to %s\n", height(trialTable), outDir);
end


function row = emptyLayer1Row()
row = struct( ...
    "trial_id", "", "layer", "", "scene_id", "", "prn_label", "", ...
    "environment", "", "center_window_id", nan, ...
    "source_interval_start_zero_based", nan, ...
    "source_interval_end_zero_based", nan, ...
    "excess_delay_truth_samples", nan, ...
    "relative_doppler_truth_hz", nan, ...
    "relative_power_truth_db", nan, "phase_truth_rad", nan, ...
    "selected_order", nan, "joint_valid", false, ...
    "injected_match", false, "delay_error_samples", nan, ...
    "doppler_error_hz", nan, "power_error_db", nan, ...
    "match_cost", nan, "joint_rss", nan, "joint_bic", nan, ...
    "snapshot_wins", nan, "failure_reason", "NOT_RUN");
end


function reason = classifyFailure(result, match)
if result.selectedOrder < 2
    reason = "NO_SECONDARY_MODEL_SELECTED";
elseif ~result.jointValid
    reason = "SELECTED_MODEL_INVALID";
elseif ~match.found
    reason = "INJECTED_PATH_OUTSIDE_TOLERANCE";
else
    reason = "PASS";
end
reason = string(reason);
end


function summary = makeLayer1Summary(trials, powers)
template = struct( ...
    "layer", "Layer1_Controlled", "relative_power_db", nan, ...
    "trial_count", nan, "recovery_count", nan, "recovery_rate", nan, ...
    "median_abs_delay_error_samples", nan, ...
    "median_abs_doppler_error_hz", nan, ...
    "median_abs_power_error_db", nan, "finite_error_count", nan);
summaryRows = repmat(template, 0, 1);
for power = powers
    subset = trials(trials.relative_power_truth_db == power, :);
    pass = subset.injected_match == 1;
    finite = isfinite(subset.delay_error_samples) ...
        & isfinite(subset.doppler_error_hz) ...
        & isfinite(subset.power_error_db);
    row = template;
    row.layer = "Layer1_Controlled";
    row.relative_power_db = power;
    row.trial_count = height(subset);
    row.recovery_count = nnz(pass);
    row.recovery_rate = nnz(pass) / max(height(subset), 1);
    if any(finite)
        row.median_abs_delay_error_samples = median(abs(subset.delay_error_samples(finite)));
        row.median_abs_doppler_error_hz = median(abs(subset.doppler_error_hz(finite)));
        row.median_abs_power_error_db = median(abs(subset.power_error_db(finite)));
    end
    row.finite_error_count = nnz(finite);
    summaryRows(end + 1, 1) = row; %#ok<AGROW>
end
summary = struct2table(summaryRows);
end
